rbac/management/management/commands/sync_schemas.py

```diff
@@ -0,0 +1,148 @@
+        entry.id = None
+        entry.pk = None
+
+    def copy_custom_principals_to_public(self, tenant):
+        """Copy custom principals from provided tenant to the public schema."""
+        principals = Principal.objects.all()
+        public_schema = Tenant.objects.get(schema_name="public")
+        for principal in principals:
+            self.stdout.write(f"Copying principal {principal.username} to public schema for tenant {tenant}.")
+            if not principal.tenant:
+                principal.tenant = tenant
+                principal.save()
+            with tenant_context(public_schema):
+                self.clear_pk(principal)
+                try:
+                    principal.save()
+                except IntegrityError as err:
+                    self.stderr.write(f"Couldn't copy principal: {principal.username}. Skipping due to:\n{err}")
+                    continue
+
+    def copy_custom_roles_to_public(self, tenant):
+        """Copy custom roles from provided tenant to the public schema."""
+        roles = Role.objects.filter(system=False)
+        public_schema = Tenant.objects.get(schema_name="public")
+        for role in roles:
+            self.stdout.write(f"Copying role {role.name} to public schema for tenant {tenant}.")
+            if not role.tenant:
+                role.tenant = tenant
+                role.save()
+            access_list = list(role.access.all())
+            with tenant_context(public_schema):
+                self.clear_pk(role)
+                try:
+                    role.save()
+                except IntegrityError as err:
+                    self.stderr.write(f"Couldn't copy role: {role.name}. Skipping due to:\n{err}")
+                    continue
+                for access in access_list:
+                    self.clear_pk(access)
+                    if not access.tenant:
+                        access.tenant = tenant
+                    access.role = role
+                    try:
+                        access.save()
+                    except IntegrityError as err:
+                        self.stderr.write(f"Couldn't copy access entry: {access}. Skipping due to:\n{err}")
+                        continue
+                    for resource_def in access.resourceDefinitions.all():
+                        self.clear_pk(resource_def)
+                        resource_def.access = access
+                        try:
+                            resource_def.save()
+                        except IntegrityError as err:
+                            self.stderr.write(f"Couldn't copy {resource_def}. Skipping due to:\n{err}")
+                            continue
+                role.access.set(access_list)
+                role.save()
+
+    def copy_custom_groups_to_public(self, tenant):
+        """Copy custom groups from provided tenant to the public schema."""
+        groups = Group.objects.filter(system=False)
```

Comment:
This will ensure we copy over custom default groups as well since we flip that flag when it's modified, correct? We have some notes in this ticket: https://issues.redhat.com/browse/RHCLOUD-11199 around handing custom default groups. I'm wondering if there will be impact with this? Presumably not yet, since we aren't serving from public. Thinking out loud, but I guess when we get to serving out of public, before we join all the roles/perms from the single default group in the public schema, we'll need a check to ensure the requested tenant doesn't have a custom default group, and if they do, serve from that vs the canonical default?

Reply:
I think you're right here, this should copy over any custom defaults, but we'll need to make sure we're using those once we switch over to serving out of public.